Tests/BinaryCompatibility.lean

```diff
@@ -0,0 +1,207 @@
+  match il.length != jl.length with
+  | true => false
+  | false => Id.run $ do
+      -- If they are, let's compare the two lists.
+      match (il, jl) with
+      | ([], []) => true
+      | (i::is, j::js) => do
+        if i ≠ j then
+          false
+        else
+          loop (ByteArray.mk $ Array.mk js) (ByteArray.mk $ Array.mk is)
+      | _ => false
+
+/- Generic Wasm module test that uses loop to check for stuff. -/
+partial def testGeneric : String → ByteArray → ByteArray → TestSeq
+  | mod, rs, os => test s!"Binary representation is compatible with {fname mod}\n=== CODE ===\n{mod}\n||| END OF CODE |||" $ loop rs os
+
+/- Here's how Main used to test a particular module:
+
+  IO.println s!"{i} is represented as:"
+  let o_parsed_module ← parseTestP moduleP i
+  match o_parsed_module.2 with
+  | .error _ => IO.println "FAIL"
+  | .ok parsed_module => do
+    IO.println s!">>> !!! >>> It is converted to bytes as: {mtob parsed_module}"
+    let f := System.mkFilePath ["/tmp", "mtob.41.wasm"]
+    let h ← IO.FS.Handle.mk f IO.FS.Mode.write
+    h.write $ mtob parsed_module
+    gO.println "It's recorded to disk at /tmp/mtob.41.wasm"
+
+Instead of this, here, we're going to:
+
+ 1. Make the reference WASM bytes with `wasm-sandbox` invocation, as provided by function `w2b`.
+ 2. Read the reference bytes into a variable.
+ 3. Compile our bytes and read those into a variable.
+ 4. Write our bytes into a file called s!"{fname}.lean".
+ 5. Byte by byte, compare the two variables, reporting errors.
+-/
+partial def moduleTestIO (x : String) : IO UInt32 := do
+  -- Run w2b against x
+  -- If we failed for any reason, we return false, otherwise, we go ahead to the next step, encoded as another function.
+  match (← w2b x) with
+  | .error _ =>
+    dbg_trace s!">>>>>>>> w2b failed!"
+    pure 35
+  | .ok _w2b_x => do
+    -- Read the reference bytes into a variable.
+    let ref_bytes ← IO.FS.Handle.mk (fname x) IO.FS.Mode.read >>= fun h => h.readBinToEnd
+    -- Compile our bytes and read those into a variable.
+    let o_parsed_module := parseP moduleP "" x
+    match o_parsed_module with
+    | .error _ =>
+      dbg_trace s!">>>>>>>> parseP failed!"
+      pure 55
+    | .ok parsed_module => do
+      dbg_trace s!">>>>>>>> parseP succeeded on {x}!"
+      let our_bytes := mtob parsed_module
+      dbg_trace s!">>>>>>>> mtob succeeded, got {our_bytes.data.toList.length} bytes!"
+      -- Write our bytes into a file called s!"{fname}.lean".
+      let h ← IO.FS.Handle.mk s!"{fname x}.lean" IO.FS.Mode.write
+      dbg_trace s!">>>>>>>> {fname x}.lean opened for writing!"
+      h.write our_bytes
+      dbg_trace s!">>>>>>>> {fname x}.lean written!"
+      dbg_trace "Running test now..."
+      lspecIO $ testGeneric x ref_bytes our_bytes
+
+def withWasmSandboxRun : IO UInt32 :=
+  -- For each element in uWasmMods, run moduleTestIO on it.
+  uWasmMods.foldlM (fun _ x => moduleTestIO x) 0
```

Comment:
did you want to use `forM`? And actually, see my other comment on `moduleTestIO`.

Reply:
No, I **hope** that UInt32 encodes error code which LSpec will pick up.
I **assume** that `0` means success.
Oh. Actually good point, this is a bug!

Reply:
The bug is that we forget everything except for the last error code. I think I made an issue for this, but I'll fix it now.